src/js/media-controller.js

```diff
@@ -447,75 +427,51 @@ const setAttr = (child, attrName, attrValue) => {
-    // which will yield an array of arrays
-    // 2. flatten that into a single array (aka map+flat aka flatMap)
-    // 3. spread this as the rest of the descendant arrays' elements to return
-    ...Array.prototype.flatMap.call(allChildNodes, getMediaUIElementDescendants)
-  ];
-};
+const traverseForMediaStateReceivers = (rootNode, mediaStateReceiverCallback) => {
+  // We (currently) don't check if descendants of the `media` (e.g. <video/>) are Media State Receivers
+  // See also: `propagateMediaState`
+  if (isMediaSlotElementDescendant(rootNode)) {
+    return;
+  }
+  // If the rootNode is a custom element that's not yet defined/ready, we don't yet know for sure
+  // whether or not it or its descendants are Media State Receivers, so wait until it's
+  // defined before attempting traversal.
+  if (isUndefinedCustomElement(rootNode)) {
+    const name = rootNode?.nodeName.toLowerCase();
+    window.customElements.whenDefined(name).then(() => {
+      // Try/traverse again once the custom element is defined
+      traverseForMediaStateReceivers(rootNode, mediaStateReceiverCallback);
+    });
+    return;
+  };
 
-const toNextMediaUIElementsState = (mutationsList = []) => {
-  const mediaChromeNodesState = mutationsList.reduce((prevMediaChromeNodesState, mutationRecord) => {
-    const { addedNodes = [], removedNodes = [], type, target, attributeName } = mutationRecord;
-    if (type === 'childList') {
-      const addedMediaChromeNodes = Array.prototype.flatMap.call(addedNodes, getMediaUIElementDescendants);
-      const removedMediaChromeNodes = Array.prototype.flatMap.call(removedNodes, getMediaUIElementDescendants);
-      prevMediaChromeNodesState.addedNodes.push(...addedMediaChromeNodes);
-      prevMediaChromeNodesState.removedNodes.push(...removedMediaChromeNodes);
-      return prevMediaChromeNodesState;
-    }
-    /** option 1, controller side */
-    if (type === 'attributes' && attributeName === MediaUIAttributes.MEDIA_CHROME_ATTRIBUTES) {
-      const attrs = getMediaUIAttributesFrom(target);
-      if (attrs.length) {
-        prevMediaChromeNodesState.addedNodes.push(target);
-      }
-      else {
-        prevMediaChromeNodesState.removedNodes.push(target);
-      }
-      return prevMediaChromeNodesState;
-    }
-    return prevMediaChromeNodesState;
-  }, {
-    addedNodes: [],
-    removedNodes: [],
-  });
-  return mediaChromeNodesState;
+  // The rootNode is itself a Media State Receiver
+  if (isMediaStateReceiver(rootNode)) {
+    mediaStateReceiverCallback(rootNode);
+  }
+
+  const { children = [] } = rootNode ?? {};
+  const shadowChildren = rootNode?.shadowRoot?.children ?? [];
+  const allChildren = [...children, ...shadowChildren];
+
+  // Traverse all children (including shadowRoot children) to see if they are/have Media State Receivers
+  allChildren.forEach(child => traverseForMediaStateReceivers(child, mediaStateReceiverCallback));
 };
 
 const propagateMediaState = (els, stateName, val) => {
```

Comment:
Answering your comment: Checking for slot=media here I don't think is needed anymore since we're checking it before adding an element as a state receiver.